Doberman/Pipeline.py

```diff
@@ -85,35 +86,39 @@ def build(self, config):
                     # all this node's requirements are created
                     node_type = kwargs.pop('type')
                     node_kwargs = {
+                            'pipeline': self,
                             'name': name,
                             'logger': self.logger,
                             '_upstream': existing_upstream} # we _ the key because of the update line below
                     node_kwargs.update(kwargs)
                     n = getattr(Doberman, node_type)(**node_kwargs)
                     setup_kwargs = {}
-                    if isinstance(n, InfluxSourceNode):
+                    if isinstance(n, Doberman.InfluxSourceNode):
                         reading_doc = self.db.get_reading_setting(name=kwargs['input_var'])
                         setup_kwargs['topic'] = reading_doc['topic']
                         setup_kwargs['config_doc'] = influx_cfg
-                    if isinstance(n, InfluxSinkNode):
+                    if isinstance(n, Doberman.InfluxSinkNode):
                         setup_kwargs['write_to_influx'] = self.db.write_to_influx
-                    if isinstance(n, BufferNode):
+                    if isinstance(n, Doberman.BufferNode):
                         setup_kwargs['strict_length'] = kwargs.get('strict', False)
-                    if isinstance(n, AlarmNode):
+                    if isinstance(n, Doberman.AlarmNode):
                         doc = self.db.get_reading_setting(name=kwargs['input_var'])
                         setup_kwargs['description'] = doc['description']
                         setup_kwargs['log_alarm'] = self.db.log_alarm
                         setup_kwargs['sensor'] = doc['sensor']
                         setup_kwargs['strict_length'] = True
-                    if isinstance(n, ControlNode):
+                        setup_kwargs['base_level'] = doc['alarm_level']
+                        setup_kwargs['escalation_config'] = alarm_cfg['escalation_config']
+                        setup_kwargs['silence_duration'] = alarm_cfg['silence_duration']
+                    if isinstance(n, Doberman.ControlNode):
                         setup_kwargs['log_command'] = self.db.log_command
                         setup_kwargs['control_target'] = kwargs['control_target']
                         setup_kwargs['control_value'] = kwargs['control_value']
                     if setup_kwargs:
                         n.setup(**kwargs)
                     n.load_config(config['node_config'].get(name, {}))
-                    self.graph.append(n)
-                    if isinstance(n, BufferNode):
+                    self.graph[n.name] = n
+                    if isinstance(n, Doberman.BufferNode):
                         num_buffer_nodes += 1
                         longest_buffer = max(longest_buffer, n.buffer.length)
```

Comment:
Isn't this a bit object-disoriented? Why not put the subclass-specific part in its own overriden build method?

Reply:
We could do that, I'll see how it looks.